### PR TITLE
Use attention_mask in pretrained transformers

### PR DESCRIPTION
**Pull request by RyujiTamaki**
Fix this issue https://github.com/allenai/allennlp/issues/3465

I also experimented with the MRPC dataset.
https://github.com/RyujiTamaki/mrpc-allennlp

Results is bellow.
```
2019-12-27 22:08:17,420 - INFO - allennlp.common.util - Metrics: {
  "best_epoch": 2,
  "peak_cpu_memory_MB": 3267.492,
  "peak_gpu_0_memory_MB": 10369,
  "peak_gpu_1_memory_MB": 19,
  "peak_gpu_2_memory_MB": 19,
  "training_duration": "0:01:17.479140",
  "training_start_epoch": 0,
  "training_epochs": 2,
  "epoch": 2,
  "training_accuracy": 0.9175637393767705,
  "training_loss": 0.20614140655274862,
  "training_cpu_memory_MB": 3267.492,
  "training_gpu_0_memory_MB": 10369,
  "training_gpu_1_memory_MB": 19,
  "training_gpu_2_memory_MB": 19,
  "validation_accuracy": 0.8505154639175257,
  "validation_loss": 0.4683948732339419,
  "best_validation_accuracy": 0.8505154639175257,
  "best_validation_loss": 0.4683948732339419
}
```

### CODE REVIEW

File: allennlp/data/token_indexers/pretrained_transformer_indexer.py [hunk 1]
```diff
@@ -31,17 +31,27 @@ class PretrainedTransformerIndexer(TokenIndexer[int]):
...
+        actual values)
     """
 
     def __init__(
-        self, model_name: str, namespace: str = "tags", token_min_padding_length: int = 0
+        self,
+        model_name: str,
+        namespace: str = "tags",
+        token_min_padding_length: int = 0,
+        mask_padding_with_zero: bool = True,
```
Comment (sai-prasanna) [inline allennlp/data/token_indexers/pretrained_transformer_indexer.py]: Any particular reason for this option?

Comment (RyujiTamaki) [reply allennlp/data/token_indexers/pretrained_transformer_indexer.py]: I implemented this pull request by referring to the following code.

https://github.com/huggingface/transformers/blob/master/src/transformers/data/processors/glue.py

When I looked for a case where mask_padding_with_zero would be False, XLNet hit it.

https://github.com/zihangdai/xlnet/blob/master/classifier_utils.py#L113-L115

I can remove this mask_padding_with_zero if I specify Input_mask. I fixed it that way because it seemed simpler.
https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_xlnet.py#L581-L587

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 2]
```diff
@@ -16,12 +16,17 @@ def __init__(self, model_name: str) -> None:
...
     @overrides
     def get_output_dim(self):
         return self.output_dim
 
-    def forward(self, token_ids: torch.LongTensor) -> torch.Tensor:  # type: ignore
+    def forward(
+        self, token_ids: torch.LongTensor, attention_mask: torch.LongTensor
+    ) -> torch.Tensor:  # type: ignore
+        if self.model_type == "xlnet":
+            return self.transformer_model(input_ids=token_ids, input_mask=attention_mask)[0]
```
Comment (matt-gardner) [inline allennlp/modules/token_embedders/pretrained_transformer_embedder.py]: Looking at the code, I'm pretty sure this will do the wrong thing - if you want to use `input_mask`, you need to compute the inverse of what you computed above.  We should just remove this case.

File: allennlp/data/token_indexers/pretrained_transformer_indexer.py [hunk 3]
```diff
@@ -91,7 +91,11 @@ def tokens_to_indices(
                     f" for the following token: {token.text}"
                 )
 
-        return {index_name: indices}
+        # The mask has 1 for real tokens and 0 for padding tokens. Only real
+        # tokens are attended to.
+        attention_mask = [1] * len(indices)
+
+        return {index_name: indices, "attention_mask": attention_mask}
```
Comment (dirkgr) [inline allennlp/data/token_indexers/pretrained_transformer_indexer.py]: At https://github.com/allenai/allennlp/blob/master/allennlp/nn/util.py#L626 (and apparently many other places), we assume that the mask is stored under the key `"mask"`.

Comment (dirkgr) [reply allennlp/data/token_indexers/pretrained_transformer_indexer.py]: We also assume it's a `LongTensor`, not a `List`.

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py (hunk 2 above)
Comment (RyujiTamaki) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py]: Thank you for your advice.
I read huggingface's xlnet implementation and seem to have misunderstood. I will delete this case.

File: allennlp/data/token_indexers/pretrained_transformer_indexer.py (hunk 3 above)
Comment (RyujiTamaki) [reply allennlp/data/token_indexers/pretrained_transformer_indexer.py]: The `tokens_to_indices` method is expected to return `Dict [str, List [int]]`, so I think it's better not to return a `LongTensor`.  Is it correct?

Comment (matt-gardner) [reply allennlp/data/token_indexers/pretrained_transformer_indexer.py]: Yes, @RyujiTamaki, you're right.  The conversion to a tensor happens in `as_padded_tensor`, not here.
